File: 3.x.x/crates-utopia/auth/src/jwt/verifier.rs
//! JWT verifier base types.

use std::collections::HashMap;

use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use serde_json::Value;

use crate::error::AuthError;
use crate::jwt::enums::{Claim, Header};

/// Configuration for JWT verification.
#[derive(Debug, Clone, Default)]
pub struct VerifierConfig {
    /// Required `iss` claim; not checked when `None`.
    pub issuer: Option<String>,
    /// Acceptable `aud` values; not checked when `None`.
    pub audience: Option<Vec<String>>,
    /// Required `typ` header; not checked when `None`.
    pub token_type: Option<String>,
    /// Skip the `exp` check when true.
    pub allow_expired: bool,
    /// Clock-skew tolerance in seconds.
    pub leeway: u64,
}

impl VerifierConfig {
    /// Create a verifier configuration.
    pub fn new() -> Self {
        Self::default()
    }

    /// Set required issuer.
    pub fn issuer(mut self, issuer: impl Into<String>) -> Self {
        self.issuer = Some(issuer.into());
        self
    }

    /// Set acceptable audience(s).
    pub fn audience(mut self, audience: impl Into<Audience>) -> Self {
        self.audience = Some(audience.into().0);
        self
    }

    /// Set required token type header.
    pub fn token_type(mut self, token_type: impl Into<String>) -> Self {
        self.token_type = Some(token_type.into());
        self
    }

    /// Allow expired tokens.
    pub fn allow_expired(mut self, allow: bool) -> Self {
        self.allow_expired = allow;
        self
    }

    /// Set clock-skew leeway in seconds.
    pub fn leeway(mut self, leeway: u64) -> Self {
        self.leeway = leeway;
        self
    }
}

/// Audience value(s) for verification.
#[derive(Debug, Clone)]
pub struct Audience(pub Vec<String>);

impl From<&str> for Audience {
    fn from(value: &str) -> Self {
        Self(vec![value.to_owned()])
    }
}

impl From<String> for Audience {
    fn from(value: String) -> Self {
        Self(vec![value])
    }
}

impl From<Vec<String>> for Audience {
    fn from(value: Vec<String>) -> Self {
        Self(value)
    }
}

/// Base functionality for compact JWS token verifiers.
pub trait Verifier {
    /// Expected JWS `alg` header value.
    fn algorithm(&self) -> &'static str;

    /// Verifier configuration.
    fn config(&self) -> &VerifierConfig;

    /// Verify the raw signature against the signing input.
    fn verify_signature(&self, signing_input: &str, signature: &[u8]) -> Result<(), AuthError>;

    /// Verify a compact JWS and return its claims.
    fn verify(&self, token: &str) -> Result<HashMap<String, Value>, AuthError> {
        let config = self.config();
        if config.leeway > i64::MAX as u64 {
            return Err(AuthError::InvalidInput("leeway cannot be negative".into()));
        }

        let segments: Vec<&str> = token.split('.').collect();
        if segments.len() != 3 {
            return Err(AuthError::Verification(
                "token must have three segments".into(),
            ));
        }

        let encoded_header = segments[0];
        let encoded_claims = segments[1];
        let encoded_signature = segments[2];

        let header = decode_segment(encoded_header, "header")?;
        let claims = decode_segment(encoded_claims, "claims")?;

        let signature = base64_url_decode(encoded_signature)
            .ok_or_else(|| AuthError::Verification("signature is not valid base64url".into()))?;

        if header.get(Header::Algorithm.as_str()) != Some(&Value::String(self.algorithm().into())) {
            return Err(AuthError::Verification("unexpected token algorithm".into()));
        }

        if let Some(expected_type) = &config.token_type {
            if header.get(Header::Type.as_str()) != Some(&Value::String(expected_type.clone())) {
                return Err(AuthError::Verification("unexpected token type".into()));
            }
        }

        self.verify_signature(&format!("{encoded_header}.{encoded_claims}"), &signature)?;

        validate_claims(&claims, config)?;

        Ok(claims)
    }
}

fn validate_claims(
    claims: &HashMap<String, Value>,
    config: &VerifierConfig,
) -> Result<(), AuthError> {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|e| AuthError::Verification(e.to_string()))?
        .as_secs() as i64;
    let leeway = i64::try_from(config.leeway).unwrap_or(i64::MAX);

    if let Some(nbf) = claims.get(Claim::NotBefore.as_str()) {
        let nbf = numeric_claim(nbf, "nbf")?;
        if now + leeway < nbf {
            return Err(AuthError::Verification("token is not yet valid".into()));
        }
    }

    if let Some(iat) = claims.get(Claim::IssuedAt.as_str()) {
        let iat = numeric_claim(iat, "iat")?;
        if now + leeway < iat {
            return Err(AuthError::Verification(
                "token was issued in the future".into(),
            ));
        }
    }

    if !config.allow_expired {
        let exp = claims
            .get(Claim::Expiration.as_str())
            .ok_or_else(|| AuthError::Verification("token is missing the \"exp\" claim".into()))?;
        let exp = numeric_claim(exp, "exp")?;
        if now >= exp + leeway {
            return Err(AuthError::Verification("token has expired".into()));
        }
    }

    if let Some(expected_issuer) = &config.issuer {
        if claims.get(Claim::Issuer.as_str()) != Some(&Value::String(expected_issuer.clone())) {
            return Err(AuthError::Verification("unexpected token issuer".into()));
        }
    }

    if let Some(expected_audiences) = &config.audience {
        let token_aud = claims.get(Claim::Audience.as_str());
        if !audience_matches(token_aud, expected_audiences) {
            return Err(AuthError::Verification("unexpected token audience".into()));
        }
    }

    Ok(())
}

fn audience_matches(token_aud: Option<&Value>, expected: &[String]) -> bool {
    let token_audiences: Vec<&str> = match token_aud {
        Some(Value::String(s)) => vec![s.as_str()],
        Some(Value::Array(items)) => items.iter().filter_map(Value::as_str).collect(),
        Some(_) | None => return false,
    };

    expected
        .iter()
        .any(|expected| token_audiences.contains(&expected.as_str()))
}

fn numeric_claim(value: &Value, name: &str) -> Result<i64, AuthError> {
    match value {
        Value::Number(number) => number
            .as_i64()
            .ok_or_else(|| AuthError::Verification(format!("invalid \"{name}\" claim"))),
        _ => Err(AuthError::Verification(format!("invalid \"{name}\" claim"))),
    }
}

fn decode_segment(segment: &str, name: &str) -> Result<HashMap<String, Value>, AuthError> {
    let label = capitalize(name);
    let decoded = base64_url_decode(segment)
        .ok_or_else(|| AuthError::Verification(format!("{label} is not valid base64url")))?;

    let value: Value = serde_json::from_slice(&decoded)
        .map_err(|_| AuthError::Verification(format!("{label} is not valid JSON")))?;

    let Value::Object(map) = value else {
        return Err(AuthError::Verification(format!(
            "{label} must be a JSON object"
        )));
    };

    Ok(map.into_iter().collect())
}

fn base64_url_decode(value: &str) -> Option<Vec<u8>> {
    URL_SAFE_NO_PAD.decode(value).ok()
}

fn capitalize(value: &str) -> String {
    let mut chars = value.chars();
    match chars.next() {
        None => String::new(),
        Some(first) => first.to_uppercase().collect::<String>() + chars.as_str(),
    }
}

File: packages/auth/src/Auth/Verifier.php
<?php

declare(strict_types=1);

namespace Utopia\Auth;

use Utopia\Auth\Enums\Claim;
use Utopia\Auth\Enums\Header;
use Utopia\Auth\Verifiers\VerificationException;

/**
 * Base class for verifying tokens carried as a compact JWS (RFC 7515).
 *
 * The mirror image of {@see Issuer}: it owns everything independent of the
 * signing algorithm — splitting the compact form, base64url/JSON decoding,
 * the "alg" guard and the standard claim checks ("exp"/"nbf"/"iat"/"iss"/"aud")
 * — and delegates the signature check itself to a subclass. The two families
 * live one level down: {@see \Utopia\Auth\Verifiers\Asymmetric} (RS256) and
 * {@see \Utopia\Auth\Verifiers\Symmetric} (HS256).
 *
 * Expectations are passed to the constructor (not fluent setters) and held
 * read-only, so a shared instance is coroutine-safe — its issuer, audience or
 * type cannot be flipped mid-verification. The issuer, audience and type are
 * only checked once supplied. Expiry is enforced by default — "exp" is required
 * and must be in the future — while "nbf"/"iat" are always enforced when
 * present; `$allowExpired` relaxes only the expiry check and `$leeway`
 * tolerates clock skew.
 */
abstract class Verifier
{
    /**
     * Acceptable "aud" values. A token passes when any of these appears in its
     * audience. When null the audience is not checked.
     *
     * @var array<int, string>|null
     */
    protected readonly ?array $audience;

    /**
     * Configuration is immutable: passed once at construction so a shared
     * instance cannot have its expectations flipped mid-verification.
     *
     * @param  string|null  $issuer  Required "iss" claim; not checked when null.
     * @param  string|array<int, string>|null  $audience  Acceptable "aud" value(s); a token passes when any appears in its audience. Not checked when null.
     * @param  string|null  $type  Required "typ" header (e.g. "at+jwt"); not checked when null, so one token kind cannot be accepted in place of another.
     * @param  bool  $allowExpired  When true, skip the "exp" check and its required-presence rule (e.g. an OIDC `id_token_hint`); "nbf"/"iat" are still enforced.
     * @param  int  $leeway  Clock-skew tolerance in seconds for the time-based claims.
     *
     * @throws \InvalidArgumentException When the leeway is negative.
     */
    public function __construct(
        protected readonly ?string $issuer = null,
        string|array|null $audience = null,
        protected readonly ?string $type = null,
        protected readonly bool $allowExpired = false,
        protected readonly int $leeway = 0,
    ) {
        if ($leeway < 0) {
            throw new \InvalidArgumentException('Leeway cannot be negative');
        }

        $this->audience = match (true) {
            $audience === null => null,
            \is_array($audience) => array_values($audience),
            default => [$audience],
        };
    }

    /**
     * The JWS "alg" header the token must carry (e.g. "RS256", "HS256").
     */
    abstract protected function getAlgorithm(): string;

    /**
     * Check the raw (binary) signature against the signing input.
     */
    abstract protected function verifySignature(string $signingInput, string $signature): bool;

    /**
     * Verify a compact JWS and return its claims.
     *
     * The signature is checked first (so claims from a forged token are never
     * trusted), then the "alg" header, then the configured claim expectations.
     *
     * @return array<string, mixed>
     *
     * @throws VerificationException When the token is malformed, the signature
     *                               is invalid, or a claim fails validation.
     */
    public function verify(string $token): array
    {
        $segments = explode('.', $token);
        if (\count($segments) !== 3) {
            throw new VerificationException('Token must have three segments');
        }

        [$encodedHeader, $encodedClaims, $encodedSignature] = $segments;

        $header = $this->decodeSegment($encodedHeader, 'header');
        $claims = $this->decodeSegment($encodedClaims, 'claims');

        $signature = $this->base64UrlDecode($encodedSignature);
        if ($signature === false) {
            throw new VerificationException('Signature is not valid base64url');
        }

        // Reject "none" and any algorithm other than ours before touching the
        // key, closing the classic algorithm-confusion hole.
        if (($header[Header::Algorithm->value] ?? null) !== $this->getAlgorithm()) {
            throw new VerificationException('Unexpected token algorithm');
        }

        if ($this->type !== null && ($header[Header::Type->value] ?? null) !== $this->type) {
            throw new VerificationException('Unexpected token type');
        }

        if (!$this->verifySignature("{$encodedHeader}.{$encodedClaims}", $signature)) {
            throw new VerificationException('Signature verification failed');
        }

        $this->validateClaims($claims);

        return $claims;
    }

    /**
     * Validate the registered claims against the configured expectations.
     *
     * @param  array<string, mixed>  $claims
     *
     * @throws VerificationException
     */
    protected function validateClaims(array $claims): void
    {
        $now = time();

        // "nbf"/"iat" bound when a token *becomes* valid; they are always
        // enforced, so a token that is not valid yet or claims a future
        // issuance is rejected even when expiry is relaxed via allowExpired().
        $nbf = $claims[Claim::NotBefore->value] ?? null;
        if ($nbf !== null) {
            if (!is_numeric($nbf)) {
                throw new VerificationException('Invalid "nbf" claim');
            }
            if ($now + $this->leeway < (int) $nbf) {
                throw new VerificationException('Token is not yet valid');
            }
        }

        $iat = $claims[Claim::IssuedAt->value] ?? null;
        if ($iat !== null) {
            if (!is_numeric($iat)) {
                throw new VerificationException('Invalid "iat" claim');
            }
            if ($now + $this->leeway < (int) $iat) {
                throw new VerificationException('Token was issued in the future');
            }
        }

        // These are bounded-lifetime bearer tokens, so "exp" is required and
        // must be in the future — unless relaxed via $allowExpired.
        if (!$this->allowExpired) {
            $exp = $claims[Claim::Expiration->value] ?? null;
            if ($exp === null) {
                throw new VerificationException('Token is missing the "exp" claim');
            }
            if (!is_numeric($exp)) {
                throw new VerificationException('Invalid "exp" claim');
            }
            if ($now >= (int) $exp + $this->leeway) {
                throw new VerificationException('Token has expired');
            }
        }

        if ($this->issuer !== null && ($claims[Claim::Issuer->value] ?? null) !== $this->issuer) {
            throw new VerificationException('Unexpected token issuer');
        }

        if ($this->audience !== null && !$this->audienceMatches($claims[Claim::Audience->value] ?? null)) {
            throw new VerificationException('Unexpected token audience');
        }
    }

    /**
     * Whether the token's "aud" claim (a string or list per RFC 7519 §4.1.3)
     * contains any of the configured acceptable audiences.
     */
    private function audienceMatches(mixed $aud): bool
    {
        $tokenAudiences = \is_array($aud) ? $aud : [$aud];

        foreach ($this->audience ?? [] as $expected) {
            if (\in_array($expected, $tokenAudiences, true)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Base64url-decode then JSON-decode a token segment into an object.
     *
     * @return array<string, mixed>
     *
     * @throws VerificationException When the segment is not base64url, not JSON,
     *                               or not a JSON object.
     */
    private function decodeSegment(string $segment, string $name): array
    {
        $label = ucfirst($name);

        $decoded = $this->base64UrlDecode($segment);
        if ($decoded === false) {
            throw new VerificationException("{$label} is not valid base64url");
        }

        try {
            $data = json_decode($decoded, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            throw new VerificationException("{$label} is not valid JSON");
        }

        // json_decode(..., true) maps both JSON objects and JSON arrays to PHP
        // arrays; a populated list means the segment was a JSON array, which is
        // not a valid JWT header/claims object.
        if (!\is_array($data) || (array_is_list($data) && $data !== [])) {
            throw new VerificationException("{$label} must be a JSON object");
        }

        /** @var array<string, mixed> $data */
        return $data;
    }

    /**
     * Base64url-decode without requiring padding (RFC 7515 §2). Returns false
     * on input outside the base64url alphabet.
     */
    protected function base64UrlDecode(string $value): string|false
    {
        return base64_decode(strtr($value, '-_', '+/'), true);
    }
}
